Prototype of a vertx HTTP server for strimzi-admin

gmcrobert opened the pull request:
This is a prototype of a suggested pluggable server using the Java
SPI. If you wish to run it, simply run a `mvn clean package`. The main 
vertx application is in the `http-server` module and it can be run using
`java -cp http-server/target/http-server-1.0-SNAPSHOT-fat.jar io.strimzi.http.server.Main`
once you have built it. 

The server has no endpoints defined and you need to add modules that implement the
`RestService` interface to the classpath. I have included 2 sample modules, `health` and
`kafka-admin`. 

The `health` module adds two endpoints as follows:
`GET /health/status`
`GET /health/liveness`

The `kafka-admin` modules adds two endpoints as follows:
`GET /kafka/topics`
`GET /kafka/topics/{topic-name}` where `{topic-name}` can be anything.

Both modules serve up static sample JSON. To run the server to include them, you simply run the following from the command line:

`java -cp kafka-admin/target/kafka-admin-1.0-SNAPSHOT-fat.jar:health/target/health-1.0-SNAPSHOT-fat.jar:http-server/target/http-server-1.0-SNAPSHOT-fat.jar io.strimzi.http.server.Main`

Signed-off-by: Graeme McRobert <mcrobert@uk.ibm.com>

gmcrobert committed:
Prototype of a vertx HTTP server for strimzi-admin
that is extensible through the Java SPI.

Signed-off-by: Graeme McRobert <mcrobert@uk.ibm.com>

tombentley commented on http-server/src/main/java/io/strimzi/http/server/RestService.java:
```diff
@@ -0,0 +1,15 @@
+ * point and a {@link io.vertx.ext.web.Router} containing the endpoints for the plugin.
+ */
+public interface RestService {
+    Future<Triplet<String, String, Router>> registerRoutes(Vertx vertx);
```
* There doesn't seem to be any benefit to using a `Triplet` over using a dedicate type (probably a class) which would allow you a change to better define the contract.
* As mentioned, the name seems to provide little value. And there's nothing stopping two modules from having the same name AFAICS, so it's not even bulletproof for the logging use case.
* By letting the module define its mount point you're making it easier to compose modules into an overall service, since you don't need to provide a mount point at the point of binding, but it's less flexible. I guess I'm OK with that, since it's simpler for now.

tombentley commented on http-server/src/main/java/io/strimzi/http/server/RestService.java:
```diff
@@ -0,0 +1,15 @@
+ * on the {@link AdminServer}. The service returns a plugin name, a mount
+ * point and a {@link io.vertx.ext.web.Router} containing the endpoints for the plugin.
+ */
+public interface RestService {
```
To me this interface is named wrong. It's defining the responsibilities of a _module_, and the module will define REST endpoints/services, so a more natural name would be `RestModule`, imho. I know you're using the java Service API, but I don't think that's enough reason to require `Service` in the name.

tombentley commented on http-server/src/main/java/io/strimzi/http/server/AdminServer.java:
```diff
@@ -0,0 +1,53 @@
+                final String mountPoint = future.result().getValue1();
+                final Router subRouter = future.result().getValue2();
+
+                router.mountSubRouter(mountPoint, subRouter);
```
IDK the contract of `mountSubRouter`, but does it provide a nice error message if you try to remount a subrouter where one has already been mounted? What about if one subrouter mounts `/foo` and another wants to mount `/foo/bar`, where `bar` is one of the endpoints defined by the first subrouter?

tombentley commented on http-server/src/main/java/io/strimzi/http/server/AdminServer.java:
```diff
@@ -0,0 +1,53 @@
+
+        return CompositeFuture.all(new ArrayList<>(modules))
+            .onSuccess(cf -> modules.forEach(future -> {
+                final String moduleName = future.result().getValue0();
```
The name seems to have very little value (you're just using it for logging).

tombentley commented on kafka-admin/src/main/java/io/strimzi/kafka/admin/KafkaAdminService.java:
```diff
@@ -0,0 +1,65 @@
+        if (ar.succeeded()) {
+            final OpenAPI3RouterFactory routerFactory = ar.result();
+
+            routerFactory.addHandlerByOperationId("ListTopics", this::listTopics);
+            routerFactory.addHandlerByOperationId("GetTopic", this::getTopic);
+
+            registerPromise.complete(Triplet.with(KafkaAdminService.class.getName(), "/kafka", routerFactory.getRouter()));
+            LOGGER.info("Endpoint '/kafka/topics' registered.");
+            LOGGER.info("Endpoint '/kafka/topics/{topic-name}' registered.");
```
This looks like a lot of boilerplate. Is there a way it could be annotation-driven?

tombentley reviewed (commented):
This isn't a bad starting point, but I'd like to understand more about the bigger picture: 

* What modules would be necessary for a minimal UI, for example?
* Where would authentication fit into this? (This might be simply my ignorance of these vertx APIs)
* How we might handle common dependencies between modules? (I'm thinking openapi types, rather than java dependencies).

scholzj reviewed (commented):
Apart from the comments from Tom, this looks mostly good to me as a start.

samuel-hawker reviewed (commented):
I think Tom raises some valid points. This looks like a good start to me, I think that a basic flow or set of instructions of how one might add a new endpoint may be useful for continued contributions.
e.g. write how you have to add an `api-spec` yaml, how that plumbs into the code etc.

But in all a good foundation.

ppatierno commented on health/src/main/java/io/strimzi/health/HealthService.java:
```diff
@@ -0,0 +1,46 @@
+        return promise.future();
+    }
+
+    private void createRouterFactory(final AsyncResult<OpenAPI3RouterFactory> ar, final Promise<Triplet<String, String, Router>> registerPromise) {
```
I would pass the `OpenAPI3RouterFactory` instance directly. I don't think this method should handle the failure of the caller (so the `OpenAPI3RouterFactory.create` method).
Then the name ... it's going to register the routes/adding handlers not creating a router factory (which is again done by the caller); I would change the method name to something like `addHandlers` even because the `registerRoutes` comes from the interface.

ppatierno commented on kafka-admin/src/main/java/io/strimzi/kafka/admin/KafkaAdminService.java:
```diff
@@ -0,0 +1,65 @@
+        return promise.future();
+    }
+
+    private void createRouterFactory(final AsyncResult<OpenAPI3RouterFactory> ar, final Promise<Triplet<String, String, Router>> registerPromise) {
```
same comment as the other service implementation

ppatierno commented on http-server/src/main/java/io/strimzi/http/server/AdminServer.java:
```diff
@@ -0,0 +1,53 @@
+
+    private Future<Router> loadRoutes() {
+
+        final Router router = Router.router(vertx);
```
you are creating a `Router` here, mounting subrouters returned by the registered services but then on line 25, you create another `Router` adding a subrouter which is the previous router with related subrouters. Can we avoid the `Router` in the middle?

ppatierno commented:
I left some comments but it's a good starting point overall.

gmcrobert commented:
@tombentley @scholzj @samuel-hawker @ppatierno Some of the questions in this PR are probably better answered in some sort of design outline. Would it be better for me to create a proposal document in the https://github.com/strimzi/proposals repository and combine the comments from this PR and the issue https://github.com/strimzi/strimzi-kafka-operator/issues/2540 so that there is a single place for the discussion in a readable form. When we are satisfied that it is the right direction then I will update the PR to reflect the code comments and any decisions from the discussion.

ppatierno commented:
@gmcrobert +1000 from me ;-)

gmcrobert commented:
I have created the proposal: https://github.com/strimzi/proposals/pull/9

gmcrobert committed:
Refactored to use the Java JPMS support

Signed-off-by: Graeme McRobert <mcrobert@uk.ibm.com>

gmcrobert committed:
Removed redundant META-INF service definitions.

Signed-off-by: Graeme McRobert <mcrobert@uk.ibm.com>

gmcrobert committed:
Add docker build support

- Added a docker build module that will build a docker image. To
  use it, run `mvn clean package` and then run `build.sh` in the
  docker-build/ directory.
- Added support for SLF4J

Signed-off-by: Graeme McRobert <mcrobert@uk.ibm.com>

sknot-rh commented:
What is the progress of this PR?

gmcrobert commented:
@sknot-rh this PR has been superseded by https://github.com/strimzi/strimzi-admin/pull/5 . I will close this PR.
